Improve Balances

marcinbodnar opened the pull request.

vgrichina commented:
Looks pretty slick, I found some discrepancies for balances however (gonna post privately).

vgrichina commented:
Moved to draft so that we don't merge accidentally before balances bug is fixed.

marcinbodnar commented:
@vgrichina I made changes to calculations, please check if it shows proper values.

vgrichina commented on src/utils/account-with-lockup.js:
```diff
@@ -34,47 +34,53 @@ async function signAndSendTransaction(receiverId, actions) {
-            }, BASE_GAS.mul(new BN(2)))
-        }
+    
+        await transferAllFromLockup(lockupAccountId, missingAmount)
```
@Patrick1904 needs to use `this.`, i.e. `await this.transferAllFromLockup(lockupAccountId, missingAmount)`

vgrichina commented on src/utils/account-with-lockup.js:
```diff
@@ -34,47 +34,53 @@ async function signAndSendTransaction(receiverId, actions) {
+        receiver_id: this.wrappedAccount.accountId
+    }, BASE_GAS.mul(new BN(2)))
+
+    await deleteLockupAccountIfPossible(lockupAccountId)
```
this.deleteLockupAccountIfPossible

vgrichina commented on src/utils/wallet.js:
```diff
@@ -157,6 +157,11 @@ class Wallet {
     }
 
+    async transferAllFromLockup() {
+        const lockupAccountId = getLockupAccountId(this.accountId)
```
```js
const account = await this.getAccount(this.accountId);
if (account.transferAllFromLockup) {
    await account.transferAllFromLockup();
}

```

marcinbodnar commented on src/utils/account-with-lockup.js:
```diff
@@ -33,46 +33,52 @@ async function signAndSendTransaction(receiverId, actions) {
 
-    return await this.wrappedAccount.signAndSendTransaction.call(this, receiverId, actions);
+export async function transferAllFromLockup(missingAmount) {
+    const lockupAccountId = getLockupAccountId(this.accountId)
```
@patrick you need to add this condition if you want it to work on testnet:
`f (process.env.REACT_APP_USE_TESTINGLOCKUP && this.accountId.length < 64) {
        lockupAccountId = `testinglockup.${this.accountId}`
    } else {
        lockupAccountId = getLockupAccountId(this.accountId)
    }`

marcinbodnar commented:
@Patrick1904 @vgrichina 
I refactored balance functionality for Profile:
 - added balance selector that will handle calculating profile balance
 - handled additional data in allAccounts hook so it will not unnecessarily call getBalance a second time
 - refactored allAccounts hook and reducer
 - removed symbols, as we already handle status for all actions
 - now it will show Global Alert if the account not exists
So it will work faster and will be a bit simpler.

Patrick1904 commented on src/utils/account-with-lockup.js:
```diff
@@ -81,7 +76,15 @@ export async function transferAllFromLockup(missingAmount) {
-    await this.deleteLockupAccountIfPossible(lockupAccountId)
+    const lockedBalance = new BN(await this.wrappedAccount.viewFunction(lockupAccountId, 'get_locked_amount'))
+    if (lockedBalance.eq(new BN(0))) {
+        await this.wrappedAccount.functionCall(lockupAccountId, 'unselect_staking_pool', {}, BASE_GAS.mul(new BN(2)))
```
@vgrichina is it necessary to check staking/pending balances with the staking pool before unselecting? I.e. is there any risk associated with unselecting the staking pool?

vgrichina replied:
Yes, unselecting active pool would fail. 

Generally it might be ok for the scope of this method (as it also means that we cannot really transfer all from lockup).

However we'd need to handle error upstream and direct user to unstake and withdraw from staking pool.

I think it also makes sense to try unselecting staking pool after we check for whether staking pool is selected (i.e. when we know there is some pool selected).

vgrichina commented on src/utils/account-with-lockup.js:
```diff
@@ -33,46 +33,58 @@ async function signAndSendTransaction(receiverId, actions) {
+
+export async function transferAllFromLockup(missingAmount) {
+    let lockupAccountId = getLockupAccountId(this.accountId)
+    if (process.env.REACT_APP_USE_TESTINGLOCKUP && this.accountId.length < 64) {
```
@Patrick1904  It makes sense to move this check inside of `getLockupAccountId` so that you get this consistent for everywhere lockup account ID is requested.

Patrick1904 replied:
Good point!

kcole16 commented:
@Patrick1904 @marcinbodnar Noticed a few issues

Account: [onenewone.near](https://explorer.near.org/accounts/onenewone.near) (mainnet)

1) Attempted to transfer entire lockup balance (unlocked), fails on `refresh_staking_pool_balance` with `Smart contract error: No staking pool selected`. This account had no NEAR staked, and hadn't selected a staking pool.

2) "Transfer to Wallet" banner doesn't update after transferring. Need to refresh the page

3) Agree with @Patrick1904 that we should hide "Transfer for Wallet" if amount is < 0.01 NEAR, as the banner just kinda hangs with 0.000003 NEAR. cc @vgrichina  @corwinharrell

marcinbodnar commented:
@Patrick1904 
> Noticed a few issues

I guess these are UI things, and transfer tokens that you worked on. Let me know if you want me to handle some of that.

Patrick1904 commented:
@marcinbodnar for some reason staked and pending release show the same balance for this lockup. It's on `patrick-lockup-2.testnet` `(testinglockup.patrick-lockup-2.testnet)`. It should be pending release only as nothing is currently staked.
![Screen Shot 2021-01-26 at 5 26 03 PM](https://user-images.githubusercontent.com/24921205/105929157-d468cc80-5ffb-11eb-8f64-c8670fdd0796.png)
![Screen Shot 2021-01-26 at 5 30 20 PM](https://user-images.githubusercontent.com/24921205/105929316-2c9fce80-5ffc-11eb-82c3-d7f945c24d42.png)

Patrick1904 commented on src/utils/account-with-lockup.js:
```diff
@@ -81,10 +81,15 @@ export async function transferAllFromLockup(missingAmount) {
+        const stakingPoolBalance = await this.wrappedAccount.viewFunction(lockupAccountId, 'get_known_deposited_balance')
+        if (stakingPoolBalance) {
+            throw new WalletError('Staking pool balance detected.', 'lockup.transferAllWithStakingPoolBalance')
+        }
```
@vgrichina I now check the staking pool balance before attempting to unselect the staking pool. If the lockup balance is zero, but there's a staking pool balance (staked or pending), I show the above error on the UI instructing the user to unstake and withdraw all funds from the pool in order to transfer the remaining lockup balance.

vgrichina replied:
@Patrick1904 It's best to call the pool contract directly vs calling lockup contract (which might have out of date info). See https://github.com/near/core-contracts/tree/master/staking-pool#interface

vgrichina commented on src/utils/wallet.js:
```diff
@@ -157,6 +157,13 @@ class Wallet {
         await (await this.getAccount(this.accountId)).sendMoney(receiverId, amount)
     }
 
+    async transferAllFromLockup() {
```
@Patrick1904 nit: I think it's best to just inline it in Redux action, no need to add more stuff to wallet object

kcole16 commented:
@Patrick1904 Getting this error after latest fix when trying to transfer all from onenewone.near

```
Unstake and withdraw all funds from the staking pool in order to transfer your remaining lockup balance to Wallet.
Error: Staking pool balance detected.
```

I've never staked from the lockup, but have staked from the main account.

marcinbodnar commented:
@Patrick1904 
> for some reason staked and pending release show the same balance for this lockup. It's on patrick-lockup-2.testnet (testinglockup.patrick-lockup-2.testnet). It should be pending release only as nothing is currently staked.

Fixed

I also checked balance refresh after transfer, `dispatch(getProfileBalance(accountId))` is enough to refresh balance. So it might be a different problem.

Also added minimum available to transfer amount (>= 0.01N).

Patrick1904 commented:
> @Patrick1904
> 
> > for some reason staked and pending release show the same balance for this lockup. It's on patrick-lockup-2.testnet (testinglockup.patrick-lockup-2.testnet). It should be pending release only as nothing is currently staked.
> 
> Fixed

@marcinbodnar I think there still might be an issue here, because the lockup balance is 60, 35 reserved for storage, 25 in staking pools, and still it says 35 available to transfer. I'm assuming it thinks that the storage amount is available. I think in this case, 0 should be available to transfer, and then if I withdraw all from the staking pool, 60 should be available to transfer, as it would transfer all unlocked tokens and destroy the lockup contract.

![Screen Shot 2021-01-27 at 12 30 40 PM](https://user-images.githubusercontent.com/24921205/106049864-7ab7df00-609b-11eb-8a85-c8cab2904a55.png)

marcinbodnar commented:
@Patrick1904 
>  I think there still might be an issue here, because the lockup balance is 60, 35 reserved for storage, 25 in staking pools, and still it says 35 available to transfer. I'm assuming it thinks that the storage amount is available. I think in this case, 0 should be available to transfer, and then if I withdraw all from the staking pool, 60 should be available to transfer, as it would transfer all unlocked tokens and destroy the lockup contract.

Fixed.

@vgrichina for this issue that @Patrick1904 mention I changed the calculation for `liquidOwnersBalance`, I tested for all the scenarios and it seems to work well, but please check the code to be sure:
https://github.com/near/near-wallet/pull/1365/commits/3ca0d7aa4500112a1e04d362123138e60e6247d2

Patrick1904 commented:
> @Patrick1904 Getting this error after latest fix when trying to transfer all from onenewone.near
> 
> ```
> Unstake and withdraw all funds from the staking pool in order to transfer your remaining lockup balance to Wallet.
> Error: Staking pool balance detected.
> ```
> 
> I've never staked from the lockup, but have staked from the main account.

@kcole16 this should be fixed now:) Sorry for repeated attempts. Please let me know if you still have issues

marcinbodnar commented:
@Patrick1904 @vgrichina calculations of total balance are ready.
